### PR TITLE
Support cgroups limits on cpu count in multiprocess and threaded scheduler

### PR DESCRIPTION
**Pull request by AlbertDeFusco**
This code has been borrowed from dask.distributed and tested in Linux docker containers. 

- [x] Tests added / passed
- [x] Passes `black dask` / `flake8 dask`

### CODE REVIEW

**Comment by mrocklin** [comment]
cc @jcrist for review

**Comment by mrocklin** [comment]
I'm somewhat concerned about adding the `psutil` dependency to dask-core for this.  We've historically made the core of dask not have any dependencies so that other libraries would be comfortable depending on it.

**Comment by TomAugspurger** [comment]
I think I prefer 2. Do you have a chance to update this @AlbertDeFusco?

**Comment by AlbertDeFusco** [comment]
Option 2 works for me. I'll be able to get back to this in a few days.

File: setup.py [hunk 1]
```diff
@@ -8,14 +8,15 @@
 # NOTE: These are tested in `continuous_integration/travis/test_imports.sh` If
 # you modify these, make sure to change the corresponding line there.
 extras_require = {
-  'array': ['numpy >= 1.13.0', 'toolz >= 0.7.3'],
+  'array': ['numpy >= 1.13.0', 'toolz >= 0.7.3', 'psutil >= 0.5'],
```
Comment (jrbourbeau) [inline setup.py]: I'd be more comfortable leaving `psutil` as an optional dependency for `'array'`, `'bag'`, `'dataframe'`, and `'delayed'`. It's already listed in our table of optional dependencies https://docs.dask.org/en/latest/install.html#optional-dependencies. Would you mind removing those requirements here?

File: continuous_integration/travis/test_imports.sh [hunk 2]
```diff
@@ -27,10 +27,10 @@ test_import () {
 conda create -n test-imports python=$PYTHON
 source activate test-imports
 
-(test_import "Core" "" "import dask, dask.threaded, dask.optimization") && \
-(test_import "Delayed" "toolz" "import dask.delayed") && \
-(test_import "Bag" "fsspec toolz partd cloudpickle" "import dask.bag") && \
-(test_import "Array" "toolz numpy" "import dask.array") && \
-(test_import "Dataframe" "fsspec numpy pandas toolz partd cloudpickle" "import dask.dataframe") && \
+(test_import "Core" "psutil" "import dask, dask.threaded, dask.optimization") && \
```
Comment (jrbourbeau) [inline continuous_integration/travis/test_imports.sh]: Similar comment here about the `psutil` requirement. These import checks can be removed too

File: dask/system.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+import math
+import os
+import sys
+
+HAVE_PSUTIL = False
+
+try:
```
Comment (jrbourbeau) [inline dask/system.py]: Small style comment, we tend to do optional imports like:

https://github.com/dask/dask/blob/7b4f90f0a708cbb97ebb498c78f7931d82ec99db/dask/diagnostics/tests/test_profiler.py#L15-L18

and then have a `if psutil is not None:` check later when `psutil` is needed

File: dask/tests/test_threaded.py [hunk 4]
```diff
@@ -109,6 +110,19 @@ def test_dont_spawn_too_many_threads():
...
+    before = threading.active_count()
+
+    dsk = {("x", i): (lambda: i,) for i in range(10)}
+    dsk["x"] = (sum, list(dsk))
+    for i in range(20):
+        get(dsk, "x")
+
+    after = threading.active_count()
+
+    assert after <= before + CPU_COUNT * 2
```
Comment (jrbourbeau) [inline dask/tests/test_threaded.py]: For my own understanding, why is there a factor of 2 here?

File: setup.py (hunk 1 above)
Comment (AlbertDeFusco) [reply setup.py]: done

File: dask/system.py (hunk 3 above)
Comment (AlbertDeFusco) [reply dask/system.py]: done, thanks!

File: dask/tests/test_threaded.py (hunk 4 above)
Comment (AlbertDeFusco) [reply dask/tests/test_threaded.py]: I don't know. It's a test I lifted from Distributed. @jcrist , do you know?

**Comment by mrocklin** [comment]
This looks good to me.  Merging in.  Thank you for the contribution @AlbertDeFusco !

**Comment by jrbourbeau** [comment]
Yeah, thanks for this contribution @AlbertDeFusco! (and for handling all the review comments)

**Pull request merged**